### DISCUSSION TITLE
Teensy 4.1 Y axis pins not working?

### DISCUSSION THREAD

**Discussion by JustinTimeToo** [grblHAL/core#378] [Q&A]
I have a Teensy 4.1. I have dual Y axis. X, Z and A axis running no problem but Y doesn't. I have removed the Dir and Step wire of the Y axis and paralleled them to the A axis and the Y axis moves. That means that my wiring is OK and that the problem is either in GRBLHal or in my Teensy. I used this setup to connect my machine.

[grbl-settings-backup-custom-2023-10-04.txt](https://github.com/grblHAL/core/files/12807742/grbl-settings-backup-custom-2023-10-04.txt)

![Teensy 4 1 Pin Assingment](https://github.com/grblHAL/core/assets/3255041/c23b7e36-3a09-46a7-9acb-db2d67c0c9fb)

**Comment by terjeio** [comment]
Bad Teensy? I happen to have a board and scope at hand and Y step and dir are working as they should (same pins).

**Reply by JustinTimeToo** [reply]
Could the Y reassigned to other pins? Maybe change the B axis to be the Y axis?

**Reply by terjeio** [reply]
Edit the board map file to reassign pins.

**Comment by JustinTimeToo** [comment]
Can you, please, tell me which file it is? I looked at motor_pins.h but it doesn't look like the correct file.

**Reply by terjeio** [reply]
It is one of the files [here](https://github.com/grblHAL/iMXRT1062/tree/master/grblHAL_Teensy4/src) ending in `_map.h`, which one is selected in [my_machine.h](https://github.com/grblHAL/iMXRT1062/blob/b7469ce0b1d520e14e9a7076929921919c72395c/grblHAL_Teensy4/src/my_machine.h#L24-L32). If none is selected [generic_map.h](https://github.com/grblHAL/iMXRT1062/blob/master/grblHAL_Teensy4/src/generic_map.h) is used.

**Comment by phil-barrett** [comment]
Try using the Web builder and select T41U5XBB - that mapping is the same as the pin diagram you posted.. 

One possibility - are you using any kind of level translation on your teensy pins (HCT logic or other)? are you sure it is ok? Have you verified that the signal is getting to the driver?  You can use a DMM on the Teensy pins - the direction pin is persistent after a move so you can cause it to flip for easy measurement. 

When I was first testing my hardware, I wrote a simple program that toggled every output pin once a second and then used an LED+resistor to test each pin. Quick way to verify the hardware is working.

**Reply by JustinTimeToo** [reply]
No, I don't use any level translation. I don't even know what they are. I used the .hex that was available to download.
Yes, the signal is getting to the driver, as I've stated above, when I connect the pulse and direction wires to the A axis. Both motors are running then.
I will try the DMM and see what it will show.

**Comment by JustinTimeToo** [comment]
I tried to compile in Arduino IDE but I got this error message:

_Arduino: 1.8.16 (Windows 10), TD: 1.58, Board: "Teensy 4.1, Serial, 600 MHz, Faster, US English"_

_In file included from_ _C:\Users\******\Documents\Arduino\libraries\grblHAL_Teensy4\examples\grblHAL_Teensy4_Upload\grblHAL_Teensy4_Upload.ino:26:_

_C:\Users\******\Documents\Arduino\libraries\grblHAL_Teensy4\src/grblHAL_Teensy4.h:22:10: fatal error: grbl/grbllib.h: No such file or directory_

   _22 | #include "grbl/grbllib.h"_

_compilation terminated._

_Error compiling for board Teensy 4.1._

The directory is there but the grbllib.h is not

**Reply by phil-barrett** [reply]
Go use the [web builder](http://svn.io-engineering.com:8080/). That way you can eliminate incorrect build as an issue.

Translation.  Hmmm.  The Teensy is a 3.3V device.  Most drivers are designed assuming 5V control input.  I use HCT logic to take the Teensy's 3.3V logic level to 5V. It is possible that your Y driver(s) are just not seeing the 3.3V signal?  Also, the resistors in the driver's opto circuits may be cutting the current low enough to prevent them from reliably operating at 3.3V. Try hooking up the Z driver to the Y signals.

**Reply by JustinTimeToo** [reply]
I tried the web builder but I can't figure out how to use the updated generic_map.h with it. I have reassigned the B axis pins to the Y axis.

**Reply by phil-barrett** [reply]
You don't need to use a _map.h file. Selecting a board causes the corresponding map file to be selected.  The mapping you are using is for the T41U5XBB board that you can select in the web builder. You can also look at the map file that it selected to make sure - click on the board map button. Unless you have a need to make changes to the source code, there is no need to build from source.

**Reply by terjeio** [reply]
An empty _grbl_ directory means that you have not [fetched the submodules](https://github.com/grblHAL/core/wiki/Compiling-grblHAL#getting-the-source-code) when downloading. And the Web Builder cannot be used to build with the generic map (perhaps I should allow that?), and certainly not with a modified version since it uses source code on the backend server. However, since the current version is precompiled it is likely not built with the generic map. `$I` output contains the board it was built for, an example:
```
...
[FIRMWARE:grblHAL]
[NVS STORAGE:*EEPROM 2K]
[DRIVER:MSP432]
[DRIVER VERSION:230828]
[BOARD:CNC BoosterPack]
[AUX IO:2,2,0,0]
...
```
